### ISSUE TITLE
GetProjectOptionsFromProjectFile` doesn't work for projects that reference a pcl library

### ISSUE THREAD

**Issue by 7sharp9** [fsharp/FSharp.Compiler.Service#343]
Take an executable that references Sqllite, a sample project can be obtained here: https://bugzilla.xamarin.com/attachment.cgi?id=9100

I used a gist for the output from FCS and xbuild as they clog things up here:
https://gist.github.com/7sharp9/f3728050a8a87309d1da

**Comment by 7sharp9** [comment]
Essentially all the implicit pcl references are missing so the type checker gives you invalid warnings:

![screen shot 2015-05-29 at 16 57 23](https://cloud.githubusercontent.com/assets/588746/7886736/f21a0b96-0623-11e5-968f-3489ed990ca5.png)

But everything compiles just fine.

**Comment by 7sharp9** [comment]
@rneatherway @dsyme Any ideas on this would be appreciated

In debug it looks like: `DesignTimeFacadeAssemblies` contains the 49 missing references:

![screen shot 2015-05-29 at 18 35 24](https://cloud.githubusercontent.com/assets/588746/7888581/7e7e4202-0631-11e5-9f95-da4f27f4a964.png)

**Comment by 7sharp9** [comment]
But should the tasks we pass into msbuild be automagically producing the correct list of references?

**Comment by 7sharp9** [comment]
Just a note, unfortunately adding assembly redirects doesn't magically fix this issue its still needs resolving, not that I thought it would fix this one.

Theres a distinct lack of detail on how portable stuff resolves references, I guess we just need to reverse engineer the msbuild targets to mach sure we are calling all the correct tasks and collecting the correct properties.  

@rneatherway Do you have any notes on this while writing this feature?

**Comment by rneatherway** [comment]
I notice in your output there is: `Target named 'ImplicitlyExpandTargetFramework' not found in the project.`

We add that here, and this is what should automagically produce the correct list: https://github.com/fsharp/FSharp.Compiler.Service/blob/master/src/fsharp/vs/service.fs#L2361

This seems likely to be the culprit.

**Comment by rneatherway** [comment]
That was a comment I wrote but forgot to post on Friday! I have to admit that I know very little about the portable profiles. If you could add your repro project as a test in a branch I will have a poke at it.

**Comment by 7sharp9** [comment]
`ImplicitlyExpandTargetFramework` is not in the fsproj being analysed but in the pcl that is referenced by that project.

**Comment by rneatherway** [comment]
Yes, but generally speaking we don't need the transitive closure of dependent assemblies do we?

Looking at the two logs you posted I notice a couple of strange things. First, FCS finds `Mono.Sqlite` while xbuild finds `SQLite.Net.Platform.Win32`, which I think is the PCL you depend on. So it looks like it was not the same project file that was being run on.

Secondly, there is the `ImplicitlyExpandDesignTimeFacades` task, which I expect would pull out those assemblies you found in the debugger. An alternative is to instead change our job to build the target `ResolveReferences` instead. This would possibly be more robust. I'm afraid testing those things will require rebuilding FCS though!

**Comment by 7sharp9** [comment]
If we have a project which references a pcl project, we need to include the implicit references that the pcl project includes as well.  By looking at the msbuild logs I was hoping something would come to light ...

I probably copied the wrong project output Ill check and repost.

**Comment by rneatherway** [comment]
Have you tried diffing them? It all looks good except that the `xbuild` output goes on to run `ImplicitlyExpandDesignTimeFacades`. Try the two build targets I suggested above:

1. `[| "ResolveReferences" |]`
2. `[| "ResolveAssemblyReferences"; "ImplicitlyExpandTargetFramework"; "ImplicitlyExpandDesignTimeFacades" |]`

**Comment by 7sharp9** [comment]
`ResolveReferences` seems to be mutually exclusive to the other options, or rather if `ResolveReferences` is included normal reference are not resolved to their path only the fully qualified name.

**Comment by rneatherway** [comment]
2nd level references test fails on Linux. Output includes:

```
InvalidCastException, ITask: Microsoft.Build.Framework.ITask, Microsoft.Build.Framework, Version=12.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a Task type: Microsoft.Build.Tasks.Message, Microsoft.Build.Tasks.v4.0, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Error initializing task Message: Cannot cast from source type to destination type.
Error initializing task Message: System.InvalidCastException: Cannot cast from source type to destination type.
  at Microsoft.Build.BuildEngine.BuildTask.InitializeTask () [0x00000] in <filename unknown>:0 
```

Classic old friend

**Comment by 7sharp9** [comment]
At the comment I cant find a permutation that will resolve pcl and normal references ...

**Comment by rneatherway** [comment]
For some reason, `ResolveReferences` doesn't seem to be causing `ResolveAssemblyReferences` to run.

My best so far is to remove the v12 MSBuild DLLs and use `[| "ResolveAssemblyReferences"; "ImplicitlyExpandTargetFramework"; "ImplicitlyExpandDesignTimeFacades" |]`. This combo makes all the tests pass (after adjusting Exception/Logging tests). However, it gives a warning that Framework 4.5.1 is unsupported on the `ToolsVersion12` test. It still works though:

```
Building target "ResolveAssemblyReferences" in project "/home/robin/dev/other/FSharp.Compiler.Service/tests/service/data/ToolsVersion12.fsproj" ("/usr/lib/mono/4.5/Microsoft.Common.targets").
Building target "GetFrameworkPaths" in project "/home/robin/dev/other/FSharp.Compiler.Service/tests/service/data/ToolsVersion12.fsproj" ("/usr/lib/mono/4.5/Microsoft.Common.targets"); "ResolveAssemblyReferences" depends on it.
Target GetFrameworkPaths started.
Task started.
Using task GetFrameworkPath from Microsoft.Build.Tasks.GetFrameworkPath, Microsoft.Build.Tasks.v4.0, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Task finished.
Task started.
Using task CreateProperty from Microsoft.Build.Tasks.CreateProperty, Microsoft.Build.Tasks.v4.0, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Task finished.
Task started.
Using task Warning from Microsoft.Build.Tasks.Warning, Microsoft.Build.Tasks.v4.0, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
TargetFrameworkVersion 'v4.5.1' not supported by this toolset (ToolsVersion: 4.0).
Task finished.
Target GetFrameworkPaths finished.
Done building target "GetFrameworkPaths" in project "/home/robin/dev/other/FSharp.Compiler.Service/tests/service/data/ToolsVersion12.fsproj" ("/usr/lib/mono/4.5/Microsoft.Common.targets"); "ResolveAssemblyReferences" depends on it.
Building target "GetReferenceAssemblyPaths" in project "/home/robin/dev/other/FSharp.Compiler.Service/tests/service/data/ToolsVersion12.fsproj" ("/usr/lib/mono/4.5/Microsoft.Common.targets"); "ResolveAssemblyReferences" depends on it.
Target GetReferenceAssemblyPaths started.
Task started.
Using task CreateProperty from Microsoft.Build.Tasks.CreateProperty, Microsoft.Build.Tasks.v4.0, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Task finished.
Task started.
Using task GetReferenceAssemblyPaths from Microsoft.Build.Tasks.GetReferenceAssemblyPaths, Microsoft.Build.Tasks.v4.0, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Looking for framework '.NETFramework,Version=v4.5.1' in root path '/usr/lib/mono/4.5/../xbuild-frameworks'
Found framework definition list '/usr/lib/mono/4.5/../xbuild-frameworks/.NETFramework/v4.5.1/RedistList/FrameworkList.xml' for framework '.NETFramework,Version=v4.5.1'
```

**Comment by 7sharp9** [comment]
Are you setting the defaulttoolsversion?

**Comment by 7sharp9** [comment]
And does the new sqllite test work too?  If so I think we should just go with whats currently working the best.

**Comment by 7sharp9** [comment]
@rneatherway I added you as a collaborator to my fork if you want to push the changes there?

**Comment by rneatherway** [comment]
I made three changes. 1. Switch away from v12, 2. Comment out setting of defaultToolsVersion, 3. Change build targets to the second option I suggested:

```
diff --git a/src/fsharp/FSharp.Compiler.Service/FSharp.Compiler.Service.fsproj b/src/fsharp/FSharp.Compiler.Service/FSharp.Compiler.Service.fsproj
index cfce6b5..e7dddbe 100644
--- a/src/fsharp/FSharp.Compiler.Service/FSharp.Compiler.Service.fsproj
+++ b/src/fsharp/FSharp.Compiler.Service/FSharp.Compiler.Service.fsproj
@@ -592,23 +592,11 @@
     <Reference Include="System" />
     <Reference Include="System.Core" />
     <Reference Include="System.Xml" />
-    <Reference Include="Microsoft.Build.Framework" Condition=" '$(TargetFrameworkVersion)' == 'v4.0'" />
-    <Reference Include="Microsoft.Build.Engine" Condition=" '$(TargetFrameworkVersion)' == 'v4.0'" />
-    <Reference Include="Microsoft.Build" Condition=" '$(TargetFrameworkVersion)' == 'v4.0'" />
-    <Reference Include="Microsoft.Build.Utilities.v4.0" Condition=" '$(TargetFrameworkVersion)' == 'v4.0'" />
-    <Reference Include="Microsoft.Build.Tasks.v4.0" Condition=" '$(TargetFrameworkVersion)' == 'v4.0'" />
-    
-    <Reference Include="Microsoft.Build.Framework, Version=12.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" Condition=" '$(TargetFrameworkVersion)' == 'v4.5'" >
-        <SpecificVersion>True</SpecificVersion>
-    </Reference>
-    <Reference Include="Microsoft.Build.Engine, Version=12.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" Condition=" '$(TargetFrameworkVersion)' == 'v4.5'" >
-        <SpecificVersion>True</SpecificVersion>
-    </Reference>
-    <Reference Include="Microsoft.Build, Version=12.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" Condition=" '$(TargetFrameworkVersion)' == 'v4.5'" >
-        <SpecificVersion>True</SpecificVersion>
-    </Reference>
-    <Reference Include="Microsoft.Build.Utilities.v12.0" Condition=" '$(TargetFrameworkVersion)' == 'v4.5'" />
-    <Reference Include="Microsoft.Build.Tasks.v12.0" Condition=" '$(TargetFrameworkVersion)' == 'v4.5'" />
+    <Reference Include="Microsoft.Build.Framework" />
+    <Reference Include="Microsoft.Build.Engine" />
+    <Reference Include="Microsoft.Build" />
+    <Reference Include="Microsoft.Build.Utilities.v4.0" />
+    <Reference Include="Microsoft.Build.Tasks.v4.0" />
     <Reference Include="FSharp.Core, Version=$(TargetFSharpCoreVersion), Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a">
       <Private>False</Private>
     </Reference>
diff --git a/src/fsharp/vs/service.fs b/src/fsharp/vs/service.fs
index 963d3c6..ee60ed9 100755
--- a/src/fsharp/vs/service.fs
+++ b/src/fsharp/vs/service.fs
@@ -2335,11 +2335,11 @@ type FSharpProjectFileInfo (fsprojFileName:string, ?properties, ?enableLogging)
     // Use the old API on Mono, with ToolsVersion = 12.0
     let CrackProjectUsingOldBuildAPI(fsprojFile:string) = 
         let engine = new Microsoft.Build.BuildEngine.Engine()
-#if FX_ATLEAST_45
-        engine.DefaultToolsVersion <- "12.0"
-#else
-        engine.DefaultToolsVersion <- "4.0"
-#endif
+// #if FX_ATLEAST_45
+//         engine.DefaultToolsVersion <- "12.0"
+// #else
+//         engine.DefaultToolsVersion <- "4.0"
+// #endif
 
         Option.iter (fun l -> engine.RegisterLogger(l)) logOpt
 
@@ -2358,7 +2358,7 @@ type FSharpProjectFileInfo (fsprojFileName:string, ?properties, ?enableLogging)
 
         let project = projectFromFile fsprojFile
 
-        project.Build([| "ResolveReferences" |])  |> ignore
+        project.Build([| "ResolveAssemblyReferences"; "ImplicitlyExpandTargetFramework"; "ImplicitlyExpandDesignTimeFacades" |])  |> ignore
         let directory = Path.GetDirectoryName project.FullFileName
 
         let getProp (p: Microsoft.Build.BuildEngine.Project) s = 
```

**Comment by rneatherway** [comment]
I'm just testing what happens if I use the new build API on mono.

I haven't updated the tests to pass yet, I don't have time right now.

**Comment by rneatherway** [comment]
Well that definitely didn't work.

**Comment by 7sharp9** [comment]
This is now fixed via #343
